### PR TITLE
Add AppHost details to the components article

### LINKED ISSUE #4: components-overview.md
In ```components-overview.md``` we don't really mention the orchestrator side of what someone needs to do when they set up Postgres within a service. We should probably explain both sides of the equation in more detail.

---
[Associated WorkItem - 189985](https://dev.azure.com/msft-skilling/Content/_workitems/edit/189985)

### PR DESCRIPTION
**Pull request by IEvangelist**
## Summary

- Add AppHost details to the components article
- Add snippets directory with CI-buildable example code

Fixes #4

---

#### Internal previews

| 📄 File | 🔗 Preview link |
|:--|:--|
| [docs/app-host-overview.md](https://github.com/dotnet/docs-aspire/blob/b455bd70eba3e28e452d4a754ceaced725e487bb/docs/app-host-overview.md) | [.NET Aspire orchestration overview](https://review.learn.microsoft.com/en-us/dotnet/aspire/app-host-overview?branch=pr-en-us-129) |
| [docs/components-overview.md](https://github.com/dotnet/docs-aspire/blob/b455bd70eba3e28e452d4a754ceaced725e487bb/docs/components-overview.md) | [.NET Aspire components overview](https://review.learn.microsoft.com/en-us/dotnet/aspire/components-overview?branch=pr-en-us-129) |
| [docs/database/postgresql-component.md](https://github.com/dotnet/docs-aspire/blob/b455bd70eba3e28e452d4a754ceaced725e487bb/docs/database/postgresql-component.md) | [.NET Aspire PostgreSQL component](https://review.learn.microsoft.com/en-us/dotnet/aspire/database/postgresql-component?branch=pr-en-us-129) |

### CODE REVIEW

File: docs/components-overview.md [hunk 1]
```diff
@@ -33,11 +31,13 @@ The following table lists the .NET Aspire components currently available for use
...
+
 ## Explore a sample component workflow
 
-.NET Aspire components streamline the process of consuming popular services and platforms. For example, you could use the .NET Aspire PostgreSQL component to connect to and utilize a PostgreSQL database. The database could be hosted on-prem or in a cloud service such as Azure, AWS, or GCP. The following steps demonstrate how to integrate this component into your app:
+.NET Aspire components streamline the process of consuming popular services and platforms. For example, consider the **.NET Aspire Application**  template. With this template, you get the [AppHost](app-host-overview.md) and [ServiceDefaults](service-defaults.md) projects. Imagine that you have a need for a Worker Service project to perform some database processing. You could use the [.NET Aspire PostgreSQL component](database/postgresql-component.md) to connect to and utilize a PostgreSQL database. The database could be hosted on-prem or in a cloud service such as Azure, AWS, or GCP. The following steps demonstrate how to integrate this component into your app:
```
Comment (adegeo) [inline docs/components-overview.md:38]: Is `Worker Service` really a proper noun in this context? Also, I don't think "project" is needed as this is a hypothetical describing a scenario, not actual implementation (yet).

I added a suggestion here, but you may want to consider "worker service" for the rest of the article where it's appropriate.

```suggestion
.NET Aspire components streamline the process of consuming popular services and platforms. For example, consider the **.NET Aspire Application**  template. With this template, you get the [AppHost](app-host-overview.md) and [ServiceDefaults](service-defaults.md) projects. Imagine that you have a need for a worker service to perform some database processing. You could use the [.NET Aspire PostgreSQL component](database/postgresql-component.md) to connect to and utilize a PostgreSQL database. The database could be hosted on-prem or in a cloud service such as Azure, AWS, or GCP. The following steps demonstrate how to integrate this component into your app:
```

File: docs/components-overview.md [hunk 2]
```diff
@@ -48,42 +48,44 @@ The following table lists the .NET Aspire components currently available for use
...
+1. In the _Program.cs_ file of your Worker Service project, call the <xref:Microsoft.Extensions.Hosting.AspirePostgreSqlNpgsqlExtensions.AddNpgsqlDataSource%2A> extension method to register `NpgsqlDataSource` as a service.
 
-    ```csharp
-    builder.AddNpgsqlDataSource("PostgreSqlConnection");
-    ```
+    :::code source="snippets/components/AspireApp/WorkerService/Program.cs" highlight="5":::
 
-    > [!NOTE]
-    > Components that are designed to connect to Azure services also support passwordless authentication and authorization using Azure RBAC, which is the recommended approach for production apps.
+    The preceding code adds the `NpgsqlDataSource` to the dependency injection container with the connection name of `"customers"`. The connection name is later used by the orchestrator project, when expressing resource dependencies.
```
Comment (adegeo) [inline docs/components-overview.md:62]: Should there be quotes in a code block? It kind of gets the point across without them. Consider describing a normal string. You wouldn't say:

> Set your string to `"Going to the store"`

You would just do

> Set your string to `Going to the store`

or

> Set your string to "Going to the store"

Or maybe you would and I'm thinking about this wrong and it's ME who should change?!?! Tell me please 😱

File: docs/components-overview.md [hunk 3]
```diff
@@ -48,42 +48,44 @@ The following table lists the .NET Aspire components currently available for use
...
 
-1. Inject the `NpgsqlDataSource` object into your controllers or service endpoints to run commands against the database:
+    :::code language="xml" source="snippets/components/AspireApp/AspireApp.AppHost/AspireApp.AppHost.csproj" highlight="16":::
 
-    ```csharp
-    public class ExampleService(NpgsqlDataSource dataSource)
-    {
-    }
-    ```
+    After the Worker Service project is referenced in the orchestrator project, the Worker Service project has its _Program.cs_ file updated to call the `AddServiceDefaults`. For more information on service defaults, see [Service defaults](service-defaults.md).
```
Comment (adegeo) [inline docs/components-overview.md:71]:
```suggestion
    After the Worker Service project is referenced in the orchestrator project, the Worker Service project has its _Program.cs_ file updated to call the `AddServiceDefaults` method. For more information on service defaults, see [Service defaults](service-defaults.md).
```

File: docs/components-overview.md [hunk 4]
```diff
@@ -48,42 +48,44 @@ The following table lists the .NET Aspire components currently available for use
...
+      - Calls <xref:Aspire.Hosting.ResourceBuilderExtensions.WithReference%2A> to add a reference to the `database`.
+      - Calls <xref:Aspire.Hosting.ProjectResourceBuilderExtensions.WithReplicas%2A> to set the number of replicas to `3`.
+
+1. Inject the `NpgsqlDataSource` object into the `Worker` to run commands against the database:
+
+    :::code source="snippets/components/AspireApp/WorkerService/Worker.cs" highlight="7,13":::
 
-After four steps, you now have a fully configured PostgreSQL database component integrated into your app. This component also configured health checks, logging, tracing, metrics, retries, and other useful capabilities for you behind the scenes. .NET Aspire components provide various options to configure each of these features.
+After several steps, you now have a fully configured PostgreSQL database component and corresponding container with connection integrated into your app. This component also configured health checks, logging, tracing, metrics, retries, and other useful capabilities for you behind the scenes. .NET Aspire components provide various options to configure each of these features.
```
Comment (adegeo) [inline docs/components-overview.md:88]: nit: would it be better to remove that intro? We assume they've done the steps, so we don't really have to call that out. Leave it in if you feel it's right. I know the way I talk and communicate comes out in the articles I write so if this (or anything I suggest) is more in line with how you're communicating the info, skip my suggestion. 😁 I added `!` for flair and excitement!

Additionally, you can cut out tracing to lower the amount of items listed. I would assume that if you can log you can trace? Maybe I'm wrong.

```suggestion
You now have a fully configured PostgreSQL database component and corresponding container with connection integrated into your app! This component also configured health checks, logging, metrics, retries, and other useful capabilities for you behind the scenes. .NET Aspire components provide various options to configure each of these features.
```

**Review by adegeo — APPROVED** [review]

File: docs/components-overview.md (hunk 2 above)
Comment (IEvangelist) [reply docs/components-overview.md:62]: I'm not really consistent with this and I really should be. I strive to refer to things as they are, for example, when I'm referring to a code element that is a string with an exact value, I tend to call it that. It's the argument that satisfies the parameter, and its value is `"customers"`. What do you typically do?

File: docs/components-overview.md [hunk 5]
```diff
@@ -33,11 +31,13 @@ The following table lists the .NET Aspire components currently available for use
...
+.NET Aspire components streamline the process of consuming popular services and platforms. For example, consider the **.NET Aspire Application**  template. With this template, you get the [AppHost](app-host-overview.md) and [ServiceDefaults](service-defaults.md) projects. Imagine that you have a need for a Worker Service project to perform some database processing. You could use the [.NET Aspire PostgreSQL component](database/postgresql-component.md) to connect to and utilize a PostgreSQL database. The database could be hosted on-prem or in a cloud service such as Azure, AWS, or GCP. The following steps demonstrate how to integrate this component into your app:
 
-1. Install the [Aspire.Npgsql](https://www.nuget.org/packages/Aspire.Npgsql) NuGet package.
+1. In the component consuming (Worker Service) project, install the [Aspire.Npgsql](https://www.nuget.org/packages/Aspire.Npgsql) NuGet package.
```
Comment (IEvangelist) [inline docs/components-overview.md:40]:
```suggestion
1. In the component consuming (worker service) project, install the [Aspire.Npgsql](https://www.nuget.org/packages/Aspire.Npgsql) NuGet package.
```

File: docs/components-overview.md [hunk 6]
```diff
@@ -48,42 +48,44 @@ The following table lists the .NET Aspire components currently available for use
...
-                      Version="[SelectVersion]" />
+    <PackageReference Include="Aspire.Npgsql" Version="[SelectVersion]" />
     ```
 
     ---
 
     For more information, see [dotnet add package](/dotnet/core/tools/dotnet-add-package) or [Manage package dependencies in .NET applications](/dotnet/core/tools/dependencies).
 
-1. In the _Program.cs_ file of your project, call the <xref:Microsoft.Extensions.Hosting.AspirePostgreSqlNpgsqlExtensions.AddNpgsqlDataSource%2A> extension method to register a `NpgsqlDataSource` for use via the dependency injection container. The method expects a connection name parameter.
+1. In the _Program.cs_ file of your Worker Service project, call the <xref:Microsoft.Extensions.Hosting.AspirePostgreSqlNpgsqlExtensions.AddNpgsqlDataSource%2A> extension method to register `NpgsqlDataSource` as a service.
```
Comment (IEvangelist) [inline docs/components-overview.md:58]:
```suggestion
1. In the _Program.cs_ file of your worker service project, call the <xref:Microsoft.Extensions.Hosting.AspirePostgreSqlNpgsqlExtensions.AddNpgsqlDataSource%2A> extension method to register `NpgsqlDataSource` as a service.
```

File: docs/components-overview.md [hunk 7]
```diff
@@ -48,42 +48,44 @@ The following table lists the .NET Aspire components currently available for use
...
+    > Components that are designed to connect to Azure services also support passwordless authentication and authorization using [Azure RBAC](/azure/role-based-access-control/overview), which is the recommended approach for production apps.
 
-    ```json
-    {
-      "ConnectionStrings": {
-        "PostgreSqlConnection": "Host=myserver;Database=test"
-      }
-    }
-    ```
+1. In your orchestrator project (the project with the _*.AppHost_ suffix), add a reference to the Worker Service project. If you're using Visual Studio, you can use the [**Add .NET Aspire Orchestrator Support**](setup-tooling.md#add-orchestration-projects) project context menu item to add the reference automatically. The following code snippet shows the project reference of the _AspireApp.AppHost.csproj_:
```
Comment (IEvangelist) [inline docs/components-overview.md:67]:
```suggestion
1. In your orchestrator project (the project with the _*.AppHost_ suffix), add a reference to the worker service project. If you're using Visual Studio, you can use the [**Add .NET Aspire Orchestrator Support**](setup-tooling.md#add-orchestration-projects) project context menu item to add the reference automatically. The following code snippet shows the project reference of the _AspireApp.AppHost.csproj_:
```

File: docs/components-overview.md [hunk 8]
```diff
@@ -48,42 +48,44 @@ The following table lists the .NET Aspire components currently available for use
...
+    After the Worker Service project is referenced in the orchestrator project, the Worker Service project has its _Program.cs_ file updated to call the `AddServiceDefaults`. For more information on service defaults, see [Service defaults](service-defaults.md).
+
+1. In the orchestrator project, update the _Program.cs_ file with the following code:
+
+    :::code source="snippets/components/AspireApp/AspireApp.AppHost/Program.cs" highlight="3-4,6-8":::
+
+    The preceding code:
+
+    - Calls <xref:Aspire.Hosting.PostgresBuilderExtensions.AddPostgresContainer%2A> and chains a call to <xref:Aspire.Hosting.PostgresBuilderExtensions.AddDatabase%2A>, adding a PostgreSQL database container to the app model with a database named `"customers"`.
+    - Chains calls on the result of the <xref:Aspire.Hosting.ProjectResourceBuilderExtensions.AddProject%2A> from the Worker Service project:
```
Comment (IEvangelist) [inline docs/components-overview.md:80]:
```suggestion
    - Chains calls on the result of the <xref:Aspire.Hosting.ProjectResourceBuilderExtensions.AddProject%2A> from the worker service project:
```

**Pull request merged**